File: backend/app/services/notification_service.py
from sqlalchemy.orm import Session
from app.models.notification import Notification


def create_notification(
    db: Session, user_id: int, type: str, message: str,
    venue_id: int | None = None, booking_id: int | None = None, payment_id: int | None = None,
) -> Notification:
    notification = Notification(
        user_id=user_id, type=type, message=message,
        venue_id=venue_id, booking_id=booking_id, payment_id=payment_id,
    )
    db.add(notification)
    db.commit()
    db.refresh(notification)
    return notification


def get_notifications_for_user(db: Session, user_id: int, limit: int = 20):
    return (
        db.query(Notification)
        .filter(Notification.user_id == user_id)
        .order_by(Notification.created_at.desc())
        .limit(limit)
        .all()
    )

File: backend/src/services/notificationService.js
import db from '../db/index.js';
import { notificationsTable } from '../models/notificationModel.js';
import { getClients } from '../utils/sseClient.js';
import {and,eq} from 'drizzle-orm';

export default {
  createNotification: async function ({ recipientId, type, payload }) {
    const [notification] = await db
      .insert(notificationsTable)
      .values({
        recipientId,
        type,
        payload,
      })
      .returning();

    const connections = getClients(recipientId);
    if(connections) {
      //  Push to every open tab simultaneously.
      // SSE wire format is strict — must be "data: <json>\n\n"
      // The double newline signals end of message to the browser's EventSource.
      // We send the full notification object (id, type, payload, createdAt)
      // so the frontend can append it to the list without a separate fetch.
      connections.forEach(r => r.write(`data: ${JSON.stringify(notification)}\n\n`))
    }
    return notification;
  },

  getNotifications: async function (userId) {
    const notifications = await db.query.notificationsTable.findMany({
      where: and(eq(notificationsTable.recipientId, userId), eq(notificationsTable.isRead, false)),
    });
    return notifications;
  },

  markAllAsRead: async function (userId) {
    await db
      .update(notificationsTable)
      .set({ isRead: true })
      .where(
        and(
          eq(notificationsTable.recipientId, userId),
          eq(notificationsTable.isRead, false)
        )
      );
  },
};
